File: URI/Paradigmas/1487 - Six Flags/flags.cpp
#include <iostream>
#include <vector>
using namespace std;
typedef vector<int> vi;

int knapsack(int W, vi wg, vi val, int n){
    int ks[n+1][W+1];
    for(int i = 0; i <= n; i++){
        for(int w = 0; w <= W; w++){
            if(i == 0 || w == 0)ks[i][w] = 0;
            else if(wg[i-1] <= w) ks[i][w] = max(val[i-1] + ks[i][w-wg[i-1]], ks[i-1][w]);
            else ks[i][w] = ks[i-1][w];
        }
    }
    return ks[n][W];
}

int main() {
    ios_base::sync_with_stdio(false);
    int instancia = 1;
    while(1){
        int N, T;
        cin >> N >> T;
        if(N == 0 && T == 0) break;
        vi tempos;
        vi pontos;
        for(int i = 0; i < N; i++){
            int d, p;
            cin >> d >> p;
            tempos.push_back(d);
            pontos.push_back(p);
        }
        int ans = knapsack(T, tempos, pontos, N);
        cout << "Instancia " << instancia << endl;
        cout << ans << endl;
        instancia++;
        cout << endl;
    }
    return 0;
}

File: URI Online Judge/Paradigmas/1487 - Six Flags/flags.py
# -*- coding: utf-8 -*-
# Felipe G. Foschiera
# URI 1487 - Six Flags

def knapsack(W, wg, val, n):
    if ks[n-1][W] != 0: return ks[n-1][W]
    if W == 0 or n == 0: return 0
    if wg[n-1] > W:
        result = knapsack(W, wg, val, n-1)
    else:
        result = max(val[n-1] + knapsack(W-wg[n-1], wg, val, n), knapsack(W, wg, val, n-1))
    ks[n-1][W] = result
    return result

inst = 1
while True:
    N, T = map(int, input().split())
    if N == 0 and T == 0: break
    tempos, pontos = [], []
    for _ in range(N):
        d, p = map(int, input().split())
        tempos.append(d)
        pontos.append(p)
    ks = [[0 for _ in range(T+1)] for _ in range(N)]
    print("Instancia", inst)
    inst += 1
    print(knapsack(T, tempos, pontos, N))
    print()